File: ts/isAnagram.ts
function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) return false;

  const sMap = new Map<string, number>();

  for (const char of s) {
    sMap.has(char)
      ? sMap.set(char, (sMap.get(char) as number) + 1)
      : sMap.set(char, 1);
  }

  for (const char of t) {
    const count = sMap.get(char);
    if (!count || count <= 0) return false;

    sMap.set(char, count - 1);
  }

  return true;
}

export default () => {
  const s = 'rat';
  const t = 'car';
  // const s = 'anagram';
  // const t = 'nagaram';
  const result = isAnagram(s, t);
  console.log(result);
};

File: rs/src/is_anagram.rs
use std::collections::HashMap;

fn is_anagram(s: String, t: String) -> bool {
    if s.len() != t.len() {
        return false;
    }

    let mut s_map = s.chars().fold(HashMap::new(), |mut acc, c| {
        *acc.entry(c).or_insert(0) += 1;
        return acc;
    });

    return t.chars().all(|x| {
        if let Some(count) = s_map.get_mut(&x) {
            if *count == 0 {
                return false;
            }
            *count -= 1;
            return true;
        } else {
            return false;
        }
    });
}

pub fn main() {
    let s = "anagram".to_string();
    let t = "nagaram".to_string();

    let result = is_anagram(s, t);
    println!("{}", result);
}
